api/webhooks/webhooks.py

```diff
@@ -43,97 +48,199 @@ class WebhookType(enum.Enum):
+    webhook: AbstractBaseExportableWebhookModel,
     data: typing.Mapping,
 ) -> requests.models.Response:
     headers = {"content-type": "application/json"}
     json_data = json.dumps(data, sort_keys=True, cls=DjangoJSONEncoder)
     if webhook.secret:
         signature = sign_payload(json_data, key=webhook.secret)
         headers.update({FLAGSMITH_SIGNATURE_HEADER: signature})
-
     return requests.post(str(webhook.url), data=json_data, headers=headers, timeout=10)
 
 
-def _call_webhook_email_on_error(
-    webhook: WebhookModels, data: typing.Mapping, webhook_type: WebhookType
+@register_task_handler()
+def call_webhook_with_failure_mail_after_retires(
+    webhook_id: int,
+    data: typing.Mapping,
+    webhook_type: str,
+    send_failure_mail: bool = False,
+    max_retries: int = 3,
+    try_count: int = 1,
 ):
+    """
+    Call a webhook with support for sending failure emails after retries.
+
+    :param webhook_id: The ID of the webhook to be called.
+    :param data: A mapping containing the data to be sent in the webhook request.
+    :param webhook_type: The type of the webhook to be triggered.
+    :param send_failure_mail: Whether to send a failure notification email (bool, default is False).
+    :param max_retries: The maximum number of retries to attempt (int, default is 3).
+    :param try_count: Stores the current retry attempt count in scheduled tasks,
+                        not needed to be specified (int, default is 1).
+    """
+
+    if try_count > max_retries:
+        raise ValueError("try_count can't be greater than max_retries")
+
+    if webhook_type == WebhookType.ORGANISATION.value:
+        webhook = OrganisationWebhook.objects.get(id=webhook_id)
+    else:
+        webhook = Webhook.objects.get(id=webhook_id)
+
+    headers = {"content-type": "application/json"}
+    json_data = json.dumps(data, sort_keys=True, cls=DjangoJSONEncoder)
+    if webhook.secret:
+        signature = sign_payload(json_data, key=webhook.secret)
+        headers.update({FLAGSMITH_SIGNATURE_HEADER: signature})
+
     try:
-        res = _call_webhook(webhook, data)
-    except requests.exceptions.RequestException as exc:
-        send_failure_email(
-            webhook,
-            data,
-            webhook_type,
-            f"N/A ({exc.__class__.__name__})",
+        res = requests.post(
+            str(webhook.url), data=json_data, headers=headers, timeout=10
         )
+    except requests.exceptions.RequestException as exc:
+        if try_count == max_retries:
+            if send_failure_mail:
+                send_failure_email(
+                    webhook,
+                    data,
+                    webhook_type,
+                    f"N/A ({exc.__class__.__name__})",
+                )
+        else:
+            call_webhook_with_failure_mail_after_retires.delay(
+                delay_until=(
+                    timezone.now() + timezone.timedelta(seconds=2**try_count)
```

Comment:
Do you think the `2` here should be configurable or at least be a constant?

Reply:
This should be pickable from the environment, there could be a use case where some user doesn't want the backoff to be 2^x. Should I move this to `constants` file and pick the constant's value from the settings file?

Reply:
Yes, please do that.

Reply:
Done. Also moved `retries` to settings. It was defaulted to 3 earlier.